File: stdlib/js/hull/web/htmx/pagination.js
/**
 * @file hull:web:htmx:pagination
 * @module hull:web:htmx:pagination
 * @description HTMX pagination widget — htmx-attributed nav over
 * the existing hull:web:pagination page-math.
 *
 * Eliminates the ~30-line pagination partial every hypermedia
 * app writes by hand. Delegates page-math to hull:web:pagination;
 * renders nav HTML with hx-get + hx-target + optional hx-push-url
 * attrs spliced into each link.
 *
 * One helper:
 *
 *   pagination.nav(total, opts)  -- rendered <nav> HTML string
 *
 * Lua parity: same surface as hull.web.htmx.pagination. Names map
 * snake_case → camelCase.
 */

import { pagination as basePagination } from "hull:web:pagination";

const ESC = { "&": "&amp;", '"': "&quot;", "<": "&lt;", ">": "&gt;", "'": "&#39;" };
function attrEscape(s) {
    return String(s == null ? "" : s).replace(/[&"<>']/g, (c) => ESC[c]);
}

function linkAttrs(url, target, pushUrl) {
    const parts = [
        `hx-get="${attrEscape(url)}"`,
        `hx-target="${attrEscape(target)}"`,
        `hx-swap="innerHTML"`,
        `href="${attrEscape(url)}"`,  // progressive enhancement
    ];
    if (pushUrl) parts.push(`hx-push-url="true"`);
    return parts.join(" ");
}

/**
 * Render the htmx-attributed pagination nav.
 *
 * @param {number} total      Total record count (for page-math).
 * @param {object} opts
 * @param {number}  [opts.page]
 * @param {number}  [opts.perPage]
 * @param {number}  [opts.defaultPerPage]
 * @param {string}  [opts.baseUrl]
 * @param {number}  [opts.window]    Default 2.
 * @param {string}  [opts.target]    hx-target. Default "body".
 * @param {boolean} [opts.pushUrl]   Default true.
 * @returns {string}  `<nav>...</nav>` or "" when pages <= 1.
 */
function nav(total, opts) {
    opts = opts || {};
    const target  = opts.target || "body";
    const pushUrl = opts.pushUrl !== false;

    // Delegate page-math. The base render() consumes snake_case
    // opts (per_page, base_url, default_per_page) — map them in.
    const p = basePagination.render(total, {
        page:             opts.page,
        per_page:         opts.perPage,
        default_per_page: opts.defaultPerPage,
        base_url:         opts.baseUrl,
        window:           opts.window,
    });
    if (!p.show) return "";

    const parts = ['<nav class="hull-pagination" aria-label="Pagination"><ul>'];

    if (p.has_prev) {
        parts.push(`<li><a ${linkAttrs(p.prev, target, pushUrl)} rel="prev" aria-label="Previous page">&larr;</a></li>`);
    }
    for (const link of p.links) {
        if (link.ellipsis) {
            parts.push('<li aria-hidden="true">&hellip;</li>');
        } else if (link.active) {
            // link.page is an integer from basePagination today;
            // escape defensively to match the rest of the tier.
            parts.push(`<li><a aria-current="page" class="hull-pagination-active">${attrEscape(link.page)}</a></li>`);
        } else {
            parts.push(`<li><a ${linkAttrs(link.url, target, pushUrl)}>${attrEscape(link.page)}</a></li>`);
        }
    }
    if (p.has_next) {
        parts.push(`<li><a ${linkAttrs(p.next, target, pushUrl)} rel="next" aria-label="Next page">&rarr;</a></li>`);
    }
    parts.push('</ul></nav>');
    return parts.join("");
}

export const pagination = { nav };
export default pagination;

File: stdlib/lua/hull/web/htmx/pagination.lua
--- HTMX pagination widget - htmx-attributed nav over the
--- existing hull/web/pagination@1 page-math.
--
-- @module hull.web.htmx.pagination
-- @license AGPL-3.0-or-later
--
-- Eliminates the ~30-line `_pagination.html` partial every
-- hypermedia app writes by hand. Delegates page-math (page count,
-- prev/next bounds, ellipsis insertion) to hull/web/pagination
-- @1; renders the nav HTML with hx-get + hx-target + optional
-- hx-push-url attrs spliced into each link.
--
-- One helper:
--
--   pagination.nav(total, opts)  -- rendered <nav> HTML string
--
-- The existing hull/web/pagination@1 module is unchanged; callers
-- that want plain (non-htmx) page reloads keep using
-- `pagination.render(total, opts)` to get the structured table
-- and render their own markup.
--
-- Setup:
--
--     app.manifest({
--         modules = {
--             "hull/web/htmx@1",
--             "hull/web/htmx/pagination@1",
--             "hull/web/pagination@1",  -- transitive dep
--         },
--         csp = "htmx",
--     })
--
-- Usage (handler):
--
--     local hxpag = require("hull.web.htmx.pagination")
--
--     app.get("/", function(req, res)
--         local total = db.query("SELECT COUNT(*) AS n FROM assets")[1].n
--         local p = require("hull.web.pagination").from_query(req)
--         local nav = hxpag.nav(total, {
--             page             = p.page,
--             per_page         = p.per_page,
--             base_url         = "/",
--             target           = "#asset-feed",
--             push_url         = true,
--         })
--         res:html(template.render("pages/home.html", {
--             rows = ...,
--             pagination_nav = nav,
--         }))
--     end)
--
-- Template (replaces the entire 30-line _pagination.html):
--     {{ pagination_nav | raw }}

local base_pagination = require("hull.web.pagination")

local pagination = {}

-- HTML attribute-value escaping (shared across the htmx widgets).
local attr_escape = require("hull.web.htmx").escape

-- Render the attrs that go on every active page link. Same set on
-- prev / next / numbered pages so the htmx behavior is uniform.
local function link_attrs(url, target, push_url)
    local parts = {
        'hx-get="' .. attr_escape(url) .. '"',
        'hx-target="' .. attr_escape(target) .. '"',
        'hx-swap="innerHTML"',
        'href="' .. attr_escape(url) .. '"',  -- progressive enhancement
    }
    if push_url then
        parts[#parts + 1] = 'hx-push-url="true"'
    end
    return table.concat(parts, " ")
end

--- Render the htmx-attributed pagination nav.
--
-- @tparam integer total       Total record count (for page-math).
-- @tparam table opts
-- @tparam[opt] integer  opts.page             Current page (1-based).
-- @tparam[opt] integer  opts.per_page         Page size.
-- @tparam[opt] integer  opts.default_per_page Default page size for
--                                             clean URLs (URL omits
--                                             per_page when it equals
--                                             default).
-- @tparam[opt] string   opts.base_url         Base URL (no ?page).
-- @tparam[opt] integer  opts.window           Pages either side of
--                                             current to always show.
--                                             Default 2.
-- @tparam[opt] string   opts.target           hx-target for nav links.
--                                             Default "body" (forces
--                                             full re-render - callers
--                                             usually want a content
--                                             container).
-- @tparam[opt] bool     opts.push_url         Default true.
-- @treturn string  `<nav>...</nav>` or "" when pages <= 1.
function pagination.nav(total, opts)
    opts = opts or {}
    local target   = opts.target or "body"
    local push_url = opts.push_url
    if push_url == nil then push_url = true end

    -- Delegate page-math.
    local p = base_pagination.render(total, opts)
    if not p.show then return "" end

    local parts = {
        '<nav class="hull-pagination" aria-label="Pagination"><ul>',
    }

    if p.has_prev then
        parts[#parts + 1] = '<li><a ' .. link_attrs(p.prev, target, push_url)
            .. ' rel="prev" aria-label="Previous page">&larr;</a></li>'
    end

    for _, link in ipairs(p.links) do
        if link.ellipsis then
            parts[#parts + 1] = '<li aria-hidden="true">&hellip;</li>'
        elseif link.active then
            -- link.page is always an integer from base_pagination
            -- today, but escape defensively to match the rest of
            -- the tier (and survive any future change to the
            -- base module's contract).
            parts[#parts + 1] = '<li><a aria-current="page" class="hull-pagination-active">'
                .. attr_escape(link.page) .. '</a></li>'
        else
            parts[#parts + 1] = '<li><a ' .. link_attrs(link.url, target, push_url)
                .. '>' .. attr_escape(link.page) .. '</a></li>'
        end
    end

    if p.has_next then
        parts[#parts + 1] = '<li><a ' .. link_attrs(p.next, target, push_url)
            .. ' rel="next" aria-label="Next page">&rarr;</a></li>'
    end

    parts[#parts + 1] = '</ul></nav>'
    return table.concat(parts)
end

return pagination
